src/content/app/tools/blast/state/action-listeners/blastActionListeners.ts

```diff
@@ -0,0 +1,150 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { PayloadAction, ListenerEffectAPI } from '@reduxjs/toolkit';
+
+import config from 'config';
+
+import { submitBlast } from '../blast-api/blastApiSlice';
+import {
+  restoreBlastSubmissions,
+  updateJob
+} from 'src/content/app/tools/blast/state/blast-results/blastResultsSlice';
+
+import {
+  saveBlastSubmission,
+  updateSavedBlastJob
+} from 'src/content/app/tools/blast/services/blastStorageService';
+
+import { POLLING_INTERVAL } from './blastActionListenerConstants';
+
+import type { BlastSubmission } from '../blast-results/blastResultsSlice';
+import type { RootState, AppDispatch } from 'src/store';
+
+export const submitBlastListener = {
+  matcher: submitBlast.matchFulfilled,
+  effect: async (
+    action: PayloadAction<{
+      submissionId: string;
+      submission: BlastSubmission;
+    }>,
+    listenerApi: ListenerEffectAPI<RootState, AppDispatch>
+  ) => {
+    const { submissionId, submission } = action.payload;
+    const { dispatch, signal } = listenerApi;
+
+    await saveBlastSubmission(submissionId, submission);
+
+    const runningJobs = submission.results.map(({ jobId }) => ({
+      submissionId,
+      jobId
+    }));
+
+    for await (const job of pollJobStatuses({ jobs: runningJobs, signal })) {
+      dispatch(updateJob(job));
+      await updateSavedBlastJob(job);
+    }
+  }
+};
+
+export const resforeBlastSubmissionsListener = {
+  actionCreator: restoreBlastSubmissions.fulfilled,
+  effect: async (
+    action: PayloadAction<Record<string, BlastSubmission>>,
+    listenerApi: ListenerEffectAPI<RootState, AppDispatch>
+  ) => {
+    const { dispatch, signal } = listenerApi;
+
+    const runningJobs = Object.entries(action.payload).flatMap(
+      ([submissionId, submission]) =>
+        submission.results
+          .filter((job) => job.status === 'RUNNING')
+          .map(({ jobId }) => ({
+            submissionId,
+            jobId
+          }))
+    );
+
+    if (!runningJobs.length) {
+      return;
+    }
+
+    for await (const job of pollJobStatuses({ jobs: runningJobs, signal })) {
+      dispatch(updateJob(job));
+      await updateSavedBlastJob(job);
+    }
+  }
+};
+
+async function* pollJobStatuses({
+  jobs,
+  signal
+}: {
+  jobs: Array<{ submissionId: string; jobId: string }>;
+  signal: AbortSignal;
+}) {
+  let shouldRun = true;
+  signal.addEventListener('abort', () => (shouldRun = false));
+
+  while (shouldRun && jobs.length) {
+    for (const job of [...jobs]) {
+      try {
+        const { status } = await fetchJobStatus({ jobId: job.jobId, signal });
+        if (['FINISHED', 'FAILURE'].includes(status)) {
+          jobs = jobs.filter(({ jobId }) => jobId !== job.jobId);
+          yield {
+            ...job,
+            fragment: { status }
+          };
+        }
+      } catch {
+        // don't care
```

Comment:
Don't we need to show anything to the user if this fails?

Reply:
We are still polling though. From user's perspective, it is no different than if the response was that the job was still running.